### PR TITLE
feat(resharding): pass correct layout to get_incoming_receipts_for_shard

### PR DESCRIPTION
**Pull request by Longarithm**
I had some trouble with `get_incoming_receipts_for_shard` if blocks range crosses the resharding boundary.
The main issue there is that we can pass either `prev_hash` or `block_hash` as a block range start and both usecases are valid.
The compromise is to pass target shard layout externally. This also removes `let shard_layout = epoch_manager.get_shard_layout_from_prev_block(&block_hash)?;` which is cryptic.

And this in turn allows to use non-contiguous shard ids in the simple resharding v3 test.

### CODE REVIEW

File: chain/chain/src/store/mod.rs [hunk 1]
```diff
@@ -234,23 +235,23 @@ pub trait ChainStoreAccess {
             }
 
             let prev_hash = header.prev_hash();
-            let shard_layout = epoch_manager.get_shard_layout_from_prev_block(&block_hash)?;
-            let prev_shard_layout = epoch_manager.get_shard_layout_from_prev_block(prev_hash)?;
+            let shard_layout = epoch_manager.get_shard_layout_from_prev_block(prev_hash)?;
```
Comment (Trisfald) [inline chain/chain/src/store/mod.rs:238]: nit: I'd try to keep naming consistent if possible

for instance `prev_shard_layout` and `current_shard_layout` for when there are 2 layouts in the same method
`shard_layout` when it's generic

Comment (Longarithm) [reply chain/chain/src/store/mod.rs]: Right.

**Review by Trisfald — APPROVED** [review]

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -3781,9 +3785,12 @@ impl Chain {
                 // we can't use hash from the current block here yet because the incoming receipts
                 // for this block is not stored yet
                 let new_receipts = collect_receipts(incoming_receipts.get(&shard_id).unwrap());
+                let shard_layout =
+                    self.epoch_manager.get_shard_layout_from_prev_block(prev_hash)?;
```
Comment (wacban) [inline chain/chain/src/chain.rs:3789]: `get_shard_layout(epoch_id)` is an alternative (if you happen to have the epoch id handy e.g. in the block header)

File: chain/chain/src/store/mod.rs [hunk 3]
```diff
@@ -210,17 +210,18 @@ pub trait ChainStoreAccess {
...
         let _span =
-            tracing::debug_span!(target: "chain", "get_incoming_receipts_for_shard", ?shard_id, ?block_hash, last_chunk_height_included).entered();
+            tracing::debug_span!(target: "chain", "get_incoming_receipts_for_shard", ?target_shard_id, ?block_hash, last_chunk_height_included).entered();
 
         let mut ret = vec![];
 
-        let target_shard_id = shard_id;
-        let target_shard_layout = epoch_manager.get_shard_layout_from_prev_block(&block_hash)?;
+        let mut current_shard_id = target_shard_id;
+        let mut current_shard_layout = target_shard_layout.clone();
```
Comment (wacban) [inline chain/chain/src/store/mod.rs:224]: Nice, maybe do the same for block hash? `let mut current_block_hash = block_hash` ?

File: chain/chain/src/store/mod.rs [hunk 4]
```diff
@@ -234,23 +235,23 @@ pub trait ChainStoreAccess {
             }
 
             let prev_hash = header.prev_hash();
-            let shard_layout = epoch_manager.get_shard_layout_from_prev_block(&block_hash)?;
             let prev_shard_layout = epoch_manager.get_shard_layout_from_prev_block(prev_hash)?;
```
Comment (wacban) [inline chain/chain/src/store/mod.rs:238]: Not your code but this is quite confusing. It effectively gets the shard layout for the current block hash so it shouldn't be called prev.

Comment (wacban) [reply chain/chain/src/store/mod.rs:238]: This loop is structured in a funny way where some of the mutable variables are updated at the beginning (shard id and shard layout) and others at the end (block hash). I think this is what was confusing me. Again, not part of your PR so feel free to leave as is but if you find a better way to do this please go ahead!

**Review by wacban — APPROVED** [review]
LGTM

**Pull request merged**